fw/connection.h

```diff
@@ -612,6 +619,35 @@ tfw_peer_for_each_conn(TfwPeer *p, int (*cb)(TfwConn *))
 	return r;
 }
 
+#define TFW_CONN_PROCESS_RESULT(r, was_stopped, extra_cond_to_out, nskb) \
+do {									\
+	/*								\
+	 * If error occurs second time (save_error_code is not zero or	\
+	 * was_stopped is true) close connection immediately with RST.	\
+	 */								\
+	r = (r != T_OK && (save_err_code != T_OK || was_stopped)) ?	\
+		SS_BLOCK_WITH_RST : r;					\
+	if (r == T_BLOCK_WITH_FIN || r == T_BLOCK_WITH_RST) {		\
+		kfree_skb(nskb);					\
+		goto out;						\
+	} else if (r && r != T_DROP) {					\
+		if (extra_cond_to_out) {				\
+			kfree_skb(nskb);				\
+			goto out;					\
+		}		 					\
+	/*								\
+	 * In case of T_BAD or system errors we close connection	\
+	 * with tcp_shutdown() and gracefully send all pending		\
+	 * responses to client. We should continue to process		\
+	 * WINDOW_UPDATE frames so, we should decrypt all skbs,		\
+	 * not drop them.						\
+	 */								\
+		save_err_code = T_BAD;					\
+	} else if (save_err_code != T_OK) {				\
+		r = save_err_code;					\
+	}								\
+} while(0)
```

Comment:
The macro isn't really generic and treats result of different function in different way. We can reduce copy & paste in `http_frame.c` and I'm not sure about keeping such a macro just for 2, still bit different, code places. Personally I'd prefer just to reference the each other code in comment.